modules/common/daemonset/daemonset.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2020 Red Hat
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package daemonset
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/util"
+	appsv1 "k8s.io/api/apps/v1"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// NewDaemonSet returns an initialized DaemonSet
+func NewDaemonSet(
+	daemonset *appsv1.DaemonSet,
+	timeout int,
+) *DaemonSet {
+	return &DaemonSet{
+		daemonset: daemonset,
+		timeout:   time.Duration(timeout) * time.Second,
+	}
+}
+
+// CreateOrPatch - creates or patches a deployment, reconciles after Xs if object won't exist.
+func (d *DaemonSet) CreateOrPatch(
+	ctx context.Context,
+	h *helper.Helper,
+) (ctrl.Result, error) {
+	daemonset := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      d.daemonset.Name,
+			Namespace: d.daemonset.Namespace,
+		},
+	}
+
+	op, err := controllerutil.CreateOrPatch(ctx, h.GetClient(), daemonset, func() error {
+		// Daemonset selector is immutable so we set this value only if
+		// a new object is going to be created
+		if daemonset.ObjectMeta.CreationTimestamp.IsZero() {
+			daemonset.Spec.Selector = d.daemonset.Spec.Selector
```

Comment:
Would it be simpler and still correct to fill daemonset.Spec.Selector at L54 above? I think (but not 100% sure) that would mean Selector is only set at Create but not during Patch so we could remove the IsZero conditional.

Reply:
not sure, we could test that, I just took this approach initially from the controllerutil example https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/example_test.go#L44